include/circt-c/Dialect/FIRRTL.h

```diff
@@ -17,8 +17,524 @@
 extern "C" {
 #endif
 
+// NOLINTBEGIN(modernize-use-using)
+
+//===----------------------------------------------------------------------===//
+// Opaque type declarations.
+//
+// Types are exposed to C bindings as structs containing opaque pointers. They
+// are not supposed to be inspected from C. This allows the underlying
+// representation to change without affecting the API users. The use of structs
+// instead of typedefs enables some type safety as structs are not implicitly
+// convertible to each other.
+//
+// Instances of these types may or may not own the underlying object. The
+// ownership semantics is defined by how an instance of the type was obtained.
+//===----------------------------------------------------------------------===//
+
+#define DEFINE_C_API_STRUCT(name, storage)                                     \
+  struct name {                                                                \
+    storage *ptr;                                                              \
+  };                                                                           \
+  typedef struct name name
+
+DEFINE_C_API_STRUCT(FirrtlContext, void);
+
+#undef DEFINE_C_API_STRUCT
+
+typedef MlirStringRef FirrtlStringRef;
+
+/// Constructs a string reference from the pointer and length. The pointer need
+/// not reference to a null-terminated string.
+#define firrtlCreateStringRef mlirStringRefCreate
+
+/// Constructs a string reference from a null-terminated C string. Prefer
+/// `firrtlCreateStringRef` if the length of the string is known.
+#define firrtlCreateStringRefFromCString mlirStringRefCreateFromCString
+
+/// Returns true if two string references are equal, false otherwise.
+#define firrtlStringRefEqual mlirStringRefEqual
+
 MLIR_DECLARE_CAPI_DIALECT_REGISTRATION(FIRRTL, firrtl);
 
+MLIR_CAPI_EXPORTED FirrtlContext firrtlCreateContext(void);
+
+MLIR_CAPI_EXPORTED void firrtlDestroyContext(FirrtlContext ctx);
+
+typedef void (*FirrtlErrorHandler)(FirrtlStringRef message, void *userData);
+
+MLIR_CAPI_EXPORTED void firrtlSetErrorHandler(FirrtlContext ctx,
+                                              FirrtlErrorHandler handler,
+                                              void *userData);
+
+MLIR_CAPI_EXPORTED void firrtlVisitCircuit(FirrtlContext ctx,
```

Comment:
I think the name `visit` is a little misleading, because normally when I think of visitors, it is for traversing IR, rather than building up IR. Maybe these could be named like `firrtlBuildCircuit`, etc.?